admission-policy/readme.adoc

```diff
@@ -0,0 +1,342 @@
+
+= Admission Control for Kubernetes on AWS
+Doc Writer <tim@styra.com>
+v1.0, 2018-01-17
+
+Kubernetes https://kubernetes.io/docs/admin/admission-controllers/[Admission Controllers]
+perform *semantic validation* of resources during create, update, and delete operations.
+In Kubernetes 1.7+, you can use OPA to enforce custom policies without recompiling or reconfiguring
+the Kubernetes API server by leveraging
+https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks[External Admission Webhooks].
+
+For example, you could enforce any of the following policies:
+
+* All images come from an AWS repository (other than a whitelist)
+* Images are pulled from the same AWS repository in the same region as the cluster
+* Namespaces are assigned to different AWS IAM groups
+
+
+== Goals
+
+This tutorial shows how to use OPA to enforce custom policies on resources in
+Kubernetes running on AWS. For the purpose of this tutorial, you will define a policy that
+prevents users from running `kubectl exec` on "privileged" containers in the
+"production" namespace.
+
+
+== Prerequisites
+
+This tutorial requires a Kubernetes 1.7 (or later) cluster. To test Kubernetes locally,
+we recommend using [minikube](https://kubernetes.io/docs/getting-started-guides/minikube/).
+Keep in mind that External Admission Webhook support in Kubernetes is currently in **alpha**.
+
+== Overview
+
+1. Create the Kubernetes cluster with `kops` and enable the External Admission Webhooks
+2. Install the OpenPolicyAgent as an External Admission Webhook
+3. Write admission control policies with OPA
+
+
+
+== Step 1: Create Cluster with External Admission Controllers via `kops`
+
+Create a Kubernetes 1.9 cluster with `kops` using the usual commands:
+
+
+```bash
+kops create cluster ...
+kops create secret ...
+```
+
+Update the Cluster spec to enable webhooks and configure client credentials.
+
+Run `kops edit <cluster-name>` and specify:
+
+```yaml
+  kubeAPIServer:
+    admissionControl:
+      - NamespaceLifecycle
+      - LimitRanger
+      - ServiceAccount
+      - PersistentVolumeLabel
+      - DefaultStorageClass
+      - DefaultTolerationSeconds
+      - GenericAdmissionWebhook
+      - ResourceQuota
+    runtimeConfig:
+      admissionregistration.k8s.io/v1alpha1: "true"
+```
+
+Then update `kops`.
+```bash
+kops update <clustername>
+```
+
+
+== Step 2: Deploy OPA on Kubernetes
+
+=== 2.1 Generate service-certificate for OPA
+
+First, create the required OpenSSL configuration files:
+
+*server.conf*
```

Comment:
Is there a directory location for this file?

Reply:
The current directory suffices; we just the files in the current directory when running some API calls.  Will clarify.